apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/signature/SignatureParser.java

```diff
@@ -32,10 +36,60 @@
 
 public class SignatureParser {
 
+    /**
+     * If the cache reaches this size we assume that the application creates a lot of dynamic queries.
+     * In that case it's inefficient to try to cache these as they are not likely to be repeated.
+     * But we still pay the price of allocating a Map.Entry and a String for the signature.
+     */
+    private static final int DISABLE_CACHE_THRESHOLD = 512;
+    /**
+     * The cache management overhead is probably not worth it for short queries
+     */
+    private static final int QUERY_LENGTH_CACHE_LOWER_THRESHOLD = 64;
+    /**
+     * We don't want to keep alive references to huge query strings
+     */
+    private static final int QUERY_LENGTH_CACHE_UPPER_THRESHOLD = 10_000;
     private final Scanner scanner = new Scanner();
+    /**
+     * Not using weak keys as the DISABLE_CACHE_THRESHOLD may not be reached when clearing stale entries.
+     * That means that we are leaking Strings but as the size of the map is limited that should not be an issue.
+     * Another reason is that ORMs like Hibernate generate equal SQL strings for the same query but don't reuse the same string instance.
+     * When relying on weak keys, we would not leverage any caching benefits if the query string is collected.
+     */
+    @Nullable
+    private ConcurrentMap<String, String> signatureCache = new ConcurrentHashMap<>(DISABLE_CACHE_THRESHOLD);
+
+    public void querySignature(String query, StringBuilder signature, boolean preparedStatement) {
```

Comment:
I imagine that even if the cache gets full over time, it still contains the **hot** queries, so it still makes sense to do lookups, while avoiding caching of new signatures. In other words, can I think it's better to use the conditions only for deciding whether to add to the cache, but not remove the cache altogether.

Reply:
That means we'll have to keep track via an AtomicBoolean whether or not the cache is active, right? But we can probably use `AtomicBoolean#getOpaque` to check if the cache is active or not. It has weaker guarantees than volatile but it guarantees to be eventually consistent.

Reply:
Wait, I don't need an extra flag at all when not using a weak map. And we have already concluded weak maps are not a good fit because ORMs don't use the same string instances but generate new but equal strings.